File: api/src/beats/domain/ledger.py
"""The ledger: a project's last N weeks, newest first, with the balance at the
close of each — every week figure on the project page, from one read.

Pure assembly over what the service has already fetched: the project (for the
goal it resolves per Monday), the contract that governs it or None, the
absences, the region's holidays by name, the hours worked by local day, and
`today`. Nothing here reads a clock, so every row is a plain function of its
inputs, and every figure goes through the same rounding (`round_hours`) as the
week route's, so the two cannot disagree on a week they both show.

What a week carries, and when a figure is None:

    worked            Σ of the seven days, request tz, a running beat included
    days              Mon..Sun hours, each rounded on its own
    effective_goal    `Project.effective_goal` for the Monday: the term's
                      nominal hours on a governed week, else the personal
                      goal / override
    contract_expected `week_expectation`: what the contract expects after
                      holidays and absences; None when nothing time-based
                      governs the week (before the first term, an objective
                      or 0 h term, no contract), 0 when it owed nothing by
                      circumstance (a full vacation week, after `ended_on`)
    balance_end       the balance at the close of the week — `closing_balances`
                      at the week's Sunday: the opening balance plus the
                      hours worked through Sunday minus the hours expected
                      through it, the Monday-morning figure before that
                      day's work. None for the current week (it has not
                      closed; the standing has today's figure), on a week
                      with no `contract_expected` — one null rule for the
                      row, so the expectation and the balance go blank
                      together; a term starting on a Saturday leaves its
                      week before the contract on both counts, and its
                      weekend hours move the next week's close — and on any
                      project the contract does not govern. A week entirely
                      after `ended_on` gets the final balance, unchanged:
                      both sides freeze on that day, so the ledger repeats
                      the frozen figure rather than going blank — the
                      closing balance is still what the reader asked for,
                      and it is the same on every row after the end.
    notes             the absences and named public holidays on the week's
                      weekdays, by date; an absence carries `half_day`, a
                      holiday its `name`. Weekends are left out: they owe
                      nothing under any term, so a holiday or absence there
                      changed nothing.

`worked` is every hour logged in the week; the balance counts only the hours
from the contract's first day through `ended_on`. On the week a contract
starts or ends mid-week the row's worked − contract_expected is therefore not
the balance's move: the hours before the first day or after the last are
shown, and charged nothing.

`since` is the contract's first day, or None. `totals` is the balance as of
today with the two terms that move it, or None when there is no balance today
— the same rule as the week route's `balance`.
"""

from collections.abc import Iterable, Mapping
from datetime import date, timedelta

from beats.domain.contracts import (
    absences_by_day,
    balance_terms,
    closing_balances,
    owes,
    round_hours,
    term_on,
    week_expectation,
)
from beats.domain.models import (
    Absence,
    Contract,
    Ledger,
    LedgerNote,
    LedgerNoteKind,
    LedgerTotals,
    LedgerWeek,
    Project,
)

_SUNDAY = timedelta(days=6)


def assemble_ledger(
    project: Project,
    contract: Contract | None,
    absences: Iterable[Absence],
    holidays: Mapping[date, str],
    worked: Mapping[date, float],
    today: date,
    weeks: int,
) -> Ledger:
    """The last `weeks` weeks ending with the one `today` is in, newest first."""
    monday = today - timedelta(days=today.weekday())
    mondays = [monday - timedelta(weeks=n) for n in range(weeks)]
    by_day = absences_by_day(absences)
    holiday_days = set(holidays)
    expected: dict[date, float | None] = {}
    closes: dict[date, float] = {}
    totals = None
    if contract is not None:
        expected = {
            week_of: week_expectation(contract, by_day.values(), holiday_days, week_of)
            for week_of in mondays
        }
        # The current week is first and has not closed — the standing shows
        # today's balance instead — and a week with no expectation has no
        # balance to close.
        closes = closing_balances(
            contract,
            by_day.values(),
            holiday_days,
            worked,
            [week_of + _SUNDAY for week_of in mondays[1:] if expected[week_of] is not None],
        )
        if owes(term_on(contract, today)):
            terms = balance_terms(contract, by_day.values(), holiday_days, worked, today)
            totals = LedgerTotals(
                expected=round_hours(terms.expected_through),
                worked=round_hours(terms.worked),
                balance=round_hours(terms.balance),
            )
    rows = [
        _week(
            project,
            contract,
            by_day,
            holidays,
            worked,
            week_of,
            expected.get(week_of),
            closes.get(week_of + _SUNDAY),
        )
        for week_of in mondays
    ]
    return Ledger(
        weeks=rows,
        since=contract.starts_on if contract is not None else None,
        totals=totals,
    )


def _week(
    project: Project,
    contract: Contract | None,
    by_day: Mapping[date, Absence],
    holidays: Mapping[date, str],
    worked: Mapping[date, float],
    week_of: date,
    expected: float | None,
    closing: float | None,
) -> LedgerWeek:
    days = [week_of + timedelta(days=offset) for offset in range(7)]
    hours = [worked.get(day, 0.0) for day in days]
    goal, goal_type = project.effective_goal(week_of)
    notes: list[LedgerNote] = []
    if contract is not None:
        for day in days[:5]:
            if day in holidays:
                notes.append(LedgerNote(date=day, kind=LedgerNoteKind.HOLIDAY, name=holidays[day]))
            absence = by_day.get(day)
            if absence is not None:
                notes.append(
                    LedgerNote(
                        date=day, kind=LedgerNoteKind(absence.type), half_day=absence.half_day
                    )
                )
    return LedgerWeek(
        week_of=week_of,
        worked=round_hours(sum(hours)),
        days=[round_hours(h) for h in hours],
        effective_goal=goal,
        effective_goal_type=goal_type,
        effective_goal_overridden=project.goal_overridden(week_of),
        contract_expected=expected,
        balance_end=round_hours(closing) if closing is not None else None,
        notes=notes,
    )

File: ui/client/entities/project/model/ledger.ts
/**
 * The ledger as the page lays it out: the earlier weeks newest first, with a
 * rule row where the contract's terms changed, where it started and where it
 * ended, and a run of empty weeks folded into one quiet row. Pure and
 * clock-free — `todayIso` decides which week is current (never in the
 * ledger: the standing carries it) and which term is still planned.
 *
 * The copy is the mockup's (docs/project-page-mockup.html): "From Mon Aug 3,
 * 2026 · Part time 80% of 42 h = 33.6 h/wk" over "was 60% · 25.2 h · “note”",
 * "·· 3 weeks away · Jul 6 – 26 · vacation ··".
 */

import { addIsoDays, mondayOfIso } from "@/shared/lib";
import { formatHours, formatSignedHours, sortTerms, termHoursPerWeek, toPercent } from "./contract";
import type { Contract, ContractTerm, Ledger, LedgerNote, LedgerWeek, ProjectKind } from "./types";
import { dateRange, formatShortDate, longDate, sundayOf, weekNumber, weekRange } from "./weekDates";

export interface LedgerWeekRow {
	type: "week";
	week: LedgerWeek;
	/** "W36". */
	label: string;
	/** "Aug 31 – Sep 6". */
	range: string;
	/** The +/− cell — `weekDelta`. */
	delta: number | null;
}

export interface RuleRow {
	type: "rule";
	rule: "opening" | "term" | "ended";
	/** The day the rule takes effect, YYYY-MM-DD. */
	date: string;
	text: string;
	/** "was 60% · 25.2 h · “Four days a week from August”". */
	sub?: string;
	/** Dated after today: shown as a ghost above the current week. */
	planned: boolean;
}

export interface QuietRow {
	type: "quiet";
	text: string;
	/** The folded weeks, newest first. */
	weeks: LedgerWeek[];
}

export type LedgerRow = LedgerWeekRow | RuleRow | QuietRow;

/** "Tue Aug 18 · sick ½", "Mon May 25 · Whit Monday" — a week's note as the row's "why". */
export function describeLedgerNote(note: LedgerNote): string {
	const when = formatShortDate(note.date);
	if (note.kind === "holiday") return `${when} · ${note.name ?? "holiday"}`;
	return `${when} · ${note.kind}${note.halfDay ? " ½" : ""}`;
}

/** "Part time 80% of 42 h = 33.6 h/wk" — a term as its rule row states it. */
function termRule(term: ContractTerm): string {
	const weekly = termHoursPerWeek(term);
	switch (term.scheduleType) {
		case "objective":
			return "Objective · no weekly expectation";
		case "full_time":
			return weekly == null ? "Full time" : `Full time · ${formatHours(weekly)}/wk`;
		case "custom":
			return weekly == null ? "Custom" : `Custom · ${formatHours(weekly)}/wk`;
		case "part_time": {
			if (weekly == null || term.fullTimeHours == null || term.percentage == null) {
				return "Part time";
			}
			const pct = toPercent(term.percentage);
			return `Part time ${pct}% of ${formatHours(term.fullTimeHours)} = ${formatHours(weekly)}/wk`;
		}
	}
}

/** "was 60% · 25.2 h" — the term a change replaced, in its shortest form. */
function wasRule(term: ContractTerm): string {
	const weekly = termHoursPerWeek(term);
	switch (term.scheduleType) {
		case "objective":
			return "was objective";
		case "full_time":
			return weekly == null ? "was full time" : `was full time · ${formatHours(weekly)}`;
		case "custom":
			return weekly == null ? "was custom" : `was ${formatHours(weekly)}`;
		case "part_time":
			return weekly == null || term.percentage == null
				? "was part time"
				: `was ${toPercent(term.percentage)}% · ${formatHours(weekly)}`;
	}
}

/** Every rule the contract puts in the ledger, dated; placement is decided below. */
function contractRules(
	contract: Contract,
	ledger: Ledger,
	weeks: LedgerWeek[],
	todayIso: string,
): RuleRow[] {
	const terms = sortTerms(contract.terms);
	const rules: RuleRow[] = terms.map((term, i) => {
		const planned = term.effectiveFrom > todayIso;
		if (i === 0) {
			const opening = contract.openingBalanceHours;
			const forward =
				formatSignedHours(opening) === "even"
					? "nothing brought forward"
					: `brought forward ${formatSignedHours(opening)}`;
			return {
				type: "rule",
				rule: "opening",
				date: term.effectiveFrom,
				text: `Contract starts ${longDate(term.effectiveFrom)} · ${termRule(term)}${planned ? " · planned" : ""}`,
				sub: forward,
				planned,
			};
		}
		const sub = [wasRule(terms[i - 1]), term.note ? `“${term.note}”` : null]
			.filter((part): part is string => part !== null)
			.join(" · ");
		return {
			type: "rule",
			rule: "term",
			date: term.effectiveFrom,
			text: `From ${longDate(term.effectiveFrom)} · ${termRule(term)}${planned ? " · planned" : ""}`,
			sub: sub || undefined,
			planned,
		};
	});
	const endedOn = contract.endedOn;
	if (endedOn) {
		const planned = endedOn > todayIso;
		const endWeek = weeks.find((w) => w.weekOf <= endedOn && endedOn <= sundayOf(w.weekOf));
		const final = ledger.totals?.balance ?? endWeek?.balanceEnd ?? null;
		const finalText = final !== null ? ` · final balance ${formatSignedHours(final)}` : "";
		rules.push({
			type: "rule",
			rule: "ended",
			date: endedOn,
			text: planned
				? `Ends ${longDate(endedOn)} · planned`
				: `Ended ${longDate(endedOn)}${finalText}`,
			planned,
		});
	}
	return rules;
}

/**
 * A week with nothing to show: worked 0, and on a week the contract governs
 * nothing owed; elsewhere no goal and no override either — a missed goal is a
 * figure, and its cell is where the week's "No goal" override is set.
 */
function isQuiet(week: LedgerWeek): boolean {
	if (week.worked !== 0) return false;
	if (week.contractExpected !== null) return week.contractExpected === 0;
	return week.effectiveGoal === null && !week.effectiveGoalOverridden;
}

function quietRow(run: LedgerWeek[]): QuietRow {
	const newest = run[0];
	const oldest = run[run.length - 1];
	const notes = run.flatMap((w) => w.notes);
	const away = notes.length > 0 && notes.every((n) => n.kind === "vacation");
	const text = away
		? `·· ${run.length} weeks away · ${dateRange(oldest.weekOf, sundayOf(newest.weekOf))} · vacation ··`
		: `·· ${run.length} quiet weeks ··`;
	return { type: "quiet", text, weeks: run };
}

/** Consecutive week rows that are all quiet, two or more of them, become one quiet row. */
function collapseQuiet(rows: LedgerRow[]): LedgerRow[] {
	const out: LedgerRow[] = [];
	let run: LedgerWeekRow[] = [];
	const flush = () => {
		if (run.length >= 2) out.push(quietRow(run.map((row) => row.week)));
		else out.push(...run);
		run = [];
	};
	for (const row of rows) {
		if (row.type === "week" && isQuiet(row.week)) {
			run.push(row);
			continue;
		}
		flush();
		out.push(row);
	}
	flush();
	return out;
}

function roundHours(hours: number): number {
	return Math.round(hours * 100) / 100;
}

/**
 * A week's +/−: the move between its close and the one before when both are
 * in the ledger; on the contract's first close — the week holding `since`, or
 * the week after when the one holding it has no close (a weekend start) — the
 * move from the opening balance. The balance counts only the contract's days,
 * so on those weeks worked − expected is not what moved it. Else worked −
 * expected, else worked − goal; null with neither, and for a week the ledger
 * does not hold. `openingBalance` is the contract's; null off a contract.
 */
export function weekDelta(
	ledger: Ledger,
	weekOf: string,
	openingBalance: number | null,
): number | null {
	const week = ledger.weeks.find((w) => w.weekOf === weekOf);
	if (!week) return null;
	const older = ledger.weeks.find((w) => w.weekOf === addIsoDays(weekOf, -7));
	if (week.balanceEnd !== null) {
		if (older && older.balanceEnd !== null) return roundHours(week.balanceEnd - older.balanceEnd);
		const startWeek = ledger.since === null ? null : mondayOfIso(ledger.since);
		const firstClose = startWeek === weekOf || (older !== undefined && startWeek === older.weekOf);
		if (openingBalance !== null && firstClose) {
			return roundHours(week.balanceEnd - openingBalance);
		}
	}
	if (week.contractExpected !== null) return roundHours(week.worked - week.contractExpected);
	if (week.effectiveGoal !== null) return roundHours(week.worked - week.effectiveGoal);
	return null;
}

function weekRow(week: LedgerWeek, delta: number | null): LedgerWeekRow {
	return {
		type: "week",
		week,
		label: weekNumber(week.weekOf),
		range: weekRange(week.weekOf),
		delta,
	};
}

/**
 * The rows of the "Earlier weeks" panel, newest first. The current week is
 * left out. A term's rule sits under the week its date falls in (reading
 * down, everything above it is under the new terms), the opening rule
 * likewise under the contract's first week, the closing rule above the week
 * the contract ended in. A rule dated in the current week or later sits at
 * the top, a ghost when it is still to come. A rule older than the oldest
 * week shown is not shown: the ledger has not reached it yet.
 */
export function ledgerRows(
	ledger: Ledger,
	project: { kind: ProjectKind; contract?: Contract },
	todayIso: string,
): LedgerRow[] {
	const thisMonday = mondayOfIso(todayIso);
	const weeks = [...ledger.weeks]
		.filter((w) => w.weekOf < thisMonday)
		.sort((a, b) => b.weekOf.localeCompare(a.weekOf));
	const contract = project.kind === "day_job" ? project.contract : undefined;
	const rules = contract ? contractRules(contract, ledger, weeks, todayIso) : [];
	const weekOfRule = (rule: RuleRow) => mondayOfIso(rule.date);

	const rows: LedgerRow[] = [];
	// Rules from the current week on, newest first, above everything.
	rows.push(
		...rules
			.filter((r) => weekOfRule(r) >= thisMonday)
			.sort((a, b) => b.date.localeCompare(a.date)),
	);
	const opening = contract ? contract.openingBalanceHours : null;
	for (const week of weeks) {
		const inWeek = rules.filter((r) => weekOfRule(r) === week.weekOf);
		rows.push(...inWeek.filter((r) => r.rule === "ended"));
		rows.push(weekRow(week, weekDelta(ledger, week.weekOf, opening)));
		rows.push(
			...inWeek.filter((r) => r.rule !== "ended").sort((a, b) => b.date.localeCompare(a.date)),
		);
	}
	return collapseQuiet(rows);
}

function csvCell(value: string | number | null): string {
	if (value === null) return "";
	const text = String(value);
	return /[",\n\r]/.test(text) ? `"${text.replace(/"/g, '""')}"` : text;
}

/**
 * The rows as CSV: `week,expected,worked,delta,balance`, the week as its
 * Monday's ISO date, a blank where a figure is null; a side project's goal
 * stands in the expected column. Rule and quiet rows are one line each, the
 * text in the first cell.
 */
export function ledgerCsv(rows: LedgerRow[]): string {
	const lines = ["week,expected,worked,delta,balance"];
	for (const row of rows) {
		if (row.type === "week") {
			const { week } = row;
			lines.push(
				[
					week.weekOf,
					csvCell(week.contractExpected ?? week.effectiveGoal),
					csvCell(week.worked),
					csvCell(row.delta),
					csvCell(week.balanceEnd),
				].join(","),
			);
		} else {
			const text = row.type === "rule" && row.sub ? `${row.text} · ${row.sub}` : row.text;
			lines.push(`${csvCell(text)},,,,`);
		}
	}
	return `${lines.join("\n")}\n`;
}
